### DISCUSSION TITLE
Python bindings

### DISCUSSION THREAD

**Discussion by RobinSchmidt** [RobinSchmidt/RS-MET#328] [Show and tell]
i have created new repo where i want to put research-and development code that is not really supposed to clutter up the main repo. first thing, i did was to figure out, how i can create python bindings for my dsp library. took me about two days to figure it all out, but now it seems to be up and running. here's some explanation of how it works:

https://github.com/RobinSchmidt/RS-MET-Research/tree/master/Projects/rsPy

so that means, in the future, i can do some research and prototyping in python *with* access to my c++ library. all i need to do is to write some small glue-code. with boost.python, the required glue code to make a c++ library function available in python is ideally a one-liner. that will allow me to use the full stack of python visualization tools to test-drive my dsp code. and, of course, i will have access to SciPy, to combine my algorithms with those available there.

**Comment by RobinSchmidt** [comment]
hey elan - do you know of any daw or audio-editor out there that allows for python scripting? i know that reaper has its own proprietary scripting language - which is very cool - but a mainstream language with many libraries like python would be even better. on the graphics apps side - blender allows for python scripting ....but daws?

**Comment by gregjazz** [comment]
Reaper supports Python scripts, too. However, I can't seem to get some Python libraries working in it. I'm not sure if I every got SciPy working in Reaper, for example.

**Comment by RobinSchmidt** [comment]
ahaaa! nice! i think, i have to check that out!

**Comment by elanhickler** [comment]
Is this helpful for live audio dsp coding for example for creating my filters of mass destruction? Edit: probably needs a fast runtime for that, and you're talking more technical stuff that doesn't need realtime.

**Comment by RobinSchmidt** [comment]
hmm...yeah - i don't know, if it will work for realtime. i didn't even know that reaper supports python scripting.  my goal was to access my dsp library from a python programming environment and jupyter notebooks - for quick testing and plotting without having to re-compile for each little change. ...but if it works in a daw too, that would be even better!

**Comment by RobinSchmidt** [comment]
i just made it work with numpy as well. before, only basic python without numpy worked. damn - this was really frustrating. some initialization code deep down in boost.python apparently didn't do its job and left some pointers uninitialized...i have no idea, if i did something wrong or this is a bug in boost.python. i had to write my own initialization function - and now it works. phew! i can now read and write from/to numpy arrays from c++ code. this is the important datatype and bridge that we will have to work with when doing dsp in python.

**Comment by RobinSchmidt** [comment]
how cool is this?!

https://mybinder.org/v2/gh/AllenDowney/ThinkDSP/master?filepath=code%2Fchap01.ipynb

interactive content with code, plots and audio. ...if i only could get my own python library (the compiled one from my c++ code) up and running in a binder cloud environment, i could create such shareable demo content using my dsp library. it's probably complicated to get a selfmade compiled python module run in a cloud-based environment - but however - being able to create and share such content using python itself as interactive language is a cool thing already - even without access to my dsp library. and who knows - maybe in the future, i can get that to work as well... the future is bright!

edit: here is the full repo: https://github.com/AllenDowney/ThinkDSP ...for inspiration what can be done
